Fix page cache warmer test flakiness, and other things.

chrisvest opened the pull request.

chrisvest committed:
Fix flakiness in PageCacheWarmupCcIT.

The test would occasionally acquire a reference to the cluster leader,
which would then concurrently switch role to follower, and thus cause
a write transaction (creating the test data) to fail.

chrisvest committed:
Fix PageCacheWarmupCcIT.

* The test data it was creating was never actually committed, but now it is.
* It was wrongly writing the expected value, rather than the actual value, of the warmed up page count to the variable that we would later check. This made the test a tautology. We now correctly check the actual number of pages loaded into memory.

chrisvest committed:
Add a `leader-only` upstream data selection strategy.

We use it in PageCacheWarmupCcIT to ensure that the read replica test copies its files from the core we actually wrote our test data to.
Without this setting, the read replica would copy profiles from any random core, and that core might not have loaded the same amount of data into memory because it has never run any queries.

chrisvest committed:
Make it possible to set the CentralJobScheduler top-level group name.

And have the causal cluster code use this to publish and tag their cluster member identities to their threads.

chrisvest committed:
Make the Read Replica server id visible in the MemberId.toString in tests.

chrisvest committed:
Make the ReadReplica instances in ClusterRule tests respect their updateConfig calls.

chrisvest committed:
Fix a generics problem (causing javac warnings) in ReadReplica and CoreClusterMember.

chrisvest committed:
Increase the likelyhood of log pruning in various stress tests by using the `continuous` checkpoint policy.

chrisvest committed:
Fix checkstyle errors.

chrisvest committed:
Attach the page cache warmup as an availability listener.

This way, warmup will also happen after every store copy, and it will be more timely in not checking availability periodically.

andrewkerr9000 commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/pagecache/WarmupAvailabilityListener.java:
```diff
@@ -0,0 +1,118 @@
+                long elapsedMillis = NANOSECONDS.toMillis( System.nanoTime() - start );
+                monitor.warmupCompleted( elapsedMillis, pagesLoaded );
+                log.debug( "Active page cache warmup took " + Format.duration( elapsedMillis ) +
+                           " to load " + pagesLoaded + " pages." );
```
Would this timing logging be better within `PageCacheWarmer#reheat`? `PageCacheWarmerKernelExtension` is the only other place to call this and contains identical logging logic, with a different instance of `Monitor`.

chrisvest replied:
There's no logging in the kernel extension. I like that the `PageCacheWarmer` does not need to concern itself with any monitoring or logging.

andrewkerr9000 replied:
In that case this lambda, and that from line 93, could be extracted to be returned from a method.

andrewkerr9000 commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/pagecache/WarmupAvailabilityListener.java:
```diff
@@ -0,0 +1,118 @@
+                long elapsedMillis = NANOSECONDS.toMillis( System.nanoTime() - start );
+                monitor.profileCompleted( elapsedMillis, pagesInMemory );
+                log.debug( "Profiled page cache in " + Format.duration( elapsedMillis ) +
+                           ", and found " + pagesInMemory + " pages in memory." );
```
As above

andrewkerr9000 commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/pagecache/WarmupAvailabilityListener.java:
```diff
@@ -0,0 +1,118 @@
+    @Override
+    public void unavailable()
+    {
+        JobScheduler.JobHandle handle = profileHandle;
```
I don't see the benefit of this. It does make it less obvious that it is the same object having `cancel` called and being set to null below.

chrisvest replied:
I could make the null-out use `compareAndSet`, but perhaps this exercise is overkill?

andrewkerr9000 replied:
Ah right, I get it now. Maybe an `AtomicReference` could be used, but as you say maybe it would be overkill.

andrewkerr9000 commented on community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/CentralJobScheduler.java:
```diff
@@ -47,15 +48,30 @@
+        {
+            Field field = ThreadGroup.class.getDeclaredField( "name" );
+            field.setAccessible( true );
+            field.set( this, name );
```
Well, it'd be nicer to set the name on instantiation, but that would require invasive refactoring of the DI modules.

chrisvest replied:
Yeah, exactly. It struck me as a poor trade-off, because the JobScheduler is created in the PlatformModule, which is created before any edition module. So trying to get edition-specific code in before JobScheduler creation would rub against the nice hierarchy we currently have.

andrewkerr9000 replied:
It could probably be done with an extra layer of abstraction on `JobScheduler` - a factory with a name setter, that instantiates a singleton `JobScheduler` when required. Not sure we could guarantee it didn't instantiate before the name is set. Doubt it would be worth the trouble.

andrewkerr9000 commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/upstream/strategies/LeaderOnlyStrategy.java:
```diff
@@ -0,0 +1,62 @@
+import org.neo4j.helpers.Service;
+
+@Service.Implementation( UpstreamDatabaseSelectionStrategy.class )
+public class LeaderOnlyStrategy extends UpstreamDatabaseSelectionStrategy
```
Do we want this to be available to production? If so do we want to document it?

chrisvest replied:
I just needed a thing like this, and don't know the answers to these questions. This is your domain, so, you tell me :)

andrewkerr9000 replied:
Perhaps @martinfurmanski has an opinion?

RagnarW replied:
@andrewkerr9000 We actually discussed this yesterday and came to the conclusion Yes. I'll make a card on our wall to do so.

chrisvest committed:
Support asynchronous cancellation in the WarmupAvailabilityListener.

chrisvest committed:
Fix the ReadReplicaHierarchicalCatchupIT and ReadReplicaToReadReplicaCatchupIT.

The visibility semantics of the ReadReplica.setUpstreamDatabaseSelectionStrategy method had changed in previous commit, but the old behaviour has now been restored.

chrisvest committed:
Tighten up the encapsulation in ReadReplica.

chrisvest committed:
Remove the timing of page cache warmup and profiling.

This simplifies the WarmupAvailabilityListener code.
The debug logging is also removed, because it was too chatty, even for debug.

andrewkerr9000 commented:
Come to think of it, you could add a metric timer to time page cache warmup/profiling. Graphite is a much more appropriate place for such things than logs. An example is `org.neo4j.metrics.source.causalclustering.RaftMessageProcessingMetric`.

chrisvest commented:
@andrewkerr9000 We could, but I don't think this thing is important enough to have metrics.

klaren commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/pagecache/PageCacheWarmerKernelExtension.java:
```diff
@@ -67,88 +57,21 @@ public void start()
-    }
-
     @Override
     public void stop() throws Throwable
```
`start` and `stop` races here. But I'm not sure that `LifecycleAdapter` is expected to handle that, I couldn't find anything in the javadocs.

chrisvest replied:
Our life cycling is single-threaded. `LifeSupport` synchronises its transitions.

klaren commented on enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/discovery/ReadReplica.java:
```diff
@@ -211,12 +211,12 @@ public File homeDir()
     public MemberId memberId()
     {
-        return new MemberId( new UUID( serverId, 0 ) );
+        return new MemberId( new UUID( ((long) serverId) << 32, 0 ) );
```
I'm a bit curious why this is changed.

chrisvest replied:
To make the serverId show up in the MemberId.shortString. Otherwise all read-replicas look the same when debugging tests.

klaren commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/pagecache/WarmupAvailabilityListener.java:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    @Override
+    public synchronized void available()
```
It looks like we could have multiple reheats happening at the same time if we have a rapid succession of available/unavailalbe events.

1. `available()` called.
2. `startWarmup()` started, long-running.
3. `unavailable()`, cancel job does nothing with the actual running job. It only prevents starting if it wasn't already started and notifies listeners.
4. `available()` again.
5. `startWarmup()` starts again simultaniously with the old one.

chrisvest replied:
Multiple warmups is not a big deal, I think, but I suppose it could lead to multiple profile jobs running at the same time, which would be inconvenient.

chrisvest replied:
So this is actually not a problem because the `PageCacheWarmer` methods are `synchronized`. And the `WarmupAvailabilityListener` is likewise `synchronized` to guarantee that we only ever cancel the correct job.

klaren reviewed (changes requested):
A few concurrency concerns.

chrisvest commented:
@klaren Please take another look.
